source/octf/analytics/statistics/FilesystemStatistics.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include <octf/analytics/statistics/FilesystemStatistics.h>
+
+namespace octf {
+
+static constexpr auto WIF_METRIC_NAME = "write invalidation factor";
+
+FilesystemStatistics::FilesystemStatistics()
+        : m_children()
+        , m_ioStats() {}
+
+FilesystemStatistics::~FilesystemStatistics() {}
+
+FilesystemStatistics::FilesystemStatistics(const FilesystemStatistics &other)
+        : m_children(other.m_children)
+        , m_ioStats(other.m_ioStats) {}
+
+FilesystemStatistics &FilesystemStatistics::operator=(
+        const FilesystemStatistics &other) {
+    if (&other != this) {
+        m_children = other.m_children;
+        m_ioStats = other.m_ioStats;
+    }
+
+    return *this;
+}
+
+void FilesystemStatistics::count(IFileSystemViewer *viewer,
+                                 const proto::trace::ParsedEvent &event) {
+    if (event.has_io()) {
+        const auto &io = event.io();
+        if (proto::trace::Discard == io.operation()) {
+            discard(event);
+            return;
+        }
+    }
+
+    if (event.has_file()) {
+        const auto &file = event.file();
+        auto id = file.id();
+
+        auto &statistics = get(viewer, viewer->getParentId(id), event);
+        statistics.count(event);
+
+        {
+            // Update statistics by file extension
+            auto ext = viewer->getFileExtension(id);
+            if (ext != "") {
+                m_children["*." + ext].count(event);
+            }
+        }
+        {
+            // Update statistics by base name
+            auto basename = viewer->getBaseName(id);
+            if (basename != "") {
+                m_children[basename + ".*"].count(event);
+            }
+        }
+    }
+}
+
+void FilesystemStatistics::getFilesystemStatistics(
+        proto::FilesystemStatistics *statistics) const {
+    fill(statistics, "");
+}
+
+FilesystemStatistics &FilesystemStatistics::get(
```

Comment:
Please also add short comment about usage. For example:

"FilesystemStatistics are organized in a tree. To get statistics of an element knowing only it's id, we need to recursively get the parentId of id until we get the id of root element. Then we can find the requested element by going down the tree by converting the ids to names of the tree elements."